In-kernel Rust on Linux (Optional) (2nd Attempt)

blitz opened the pull request:
## Description of changes

This MR adds optional and experimental Rust support for Linux kernels. It supersedes https://github.com/NixOS/nixpkgs/pull/234311 and #271222. With this PR, you can use Rust support in Linux 6.6 and 6.7-rc2, if you opt-in to it via kernel features.

I've chosen the opt-in approach for now, because after a kernel or rustc updates, we might need to go hunt for patches and failure to find some should not block either Linux or rustc updates. But with this support, it should be vastly easier for people to experiment with Rust support already.

To make linux_testing work, we also need a bindgen update:

- https://github.com/NixOS/nixpkgs/pull/255631

Because in previous incarnations of Rust support, we had the issue that Rust support was silently broken, I've also added the official out-of-tree Rust module as a test. Now you can test everything end-to-end.

**Note to reviewers:** The original PR #271222 was closed, because I review-spammed people by accident. But there might still be discussions of interest.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

blitz committed:
rust-out-of-tree-module: init

Adding this as an example and test whether the Rust support in the
Linux kernel works.

blitz committed:
nixos/tests/kernel-rust: init

blitz committed:
doc: explain how to enable Rust support in the Linux kernel

philiptaron commented on nixos/doc/manual/configuration/linux-kernel.chapter.md:
```diff
@@ -92,6 +92,24 @@ To use your custom kernel package in your NixOS configuration, set
+## Rust
+
+The Linux kernel does not have Rust language support enabled by
+default. For kernels newer than 6.6, experimental Rust support can be
```
> For kernels newer than 6.6

Based on the test, this actually should read "For kernel versions 6.6 or newer,"

philiptaron commented on nixos/tests/kernel-rust.nix:
```diff
@@ -0,0 +1,30 @@
+      ];
+    };
+
+  testScript = ''
+    machine.wait_for_unit("default.target")
+    machine.succeed("modprobe rust_out_of_tree")
+  '';
```
Nice test.

Kloenk replied:
there are also sample drivers in tree for rust, which also get used by the rfl ci

philiptaron commented on pkgs/top-level/linux-kernels.nix:
```diff
@@ -458,6 +460,8 @@ in {
 
     facetimehd = callPackage ../os-specific/linux/facetimehd { };
 
+    rust-out-of-tree-module = if lib.versionAtLeast kernel.version "6.5" then callPackage ../os-specific/linux/rust-out-of-tree-module { } else null;
```
6.5 => 6.6 per the rest of the PR?

blitz replied:
Fixed!

philiptaron reviewed (approved):
I took a look, and will test after work hours.

philiptaron commented:
[OfBorg](https://logs.ofborg.org/logfile/nixos/nixpkgs.272029/5cc67407-6653-41e3-8172-a29317a4dab9) is unhappy, but at first inspection it's due to a Python doctest getting ConnectionReset:

```
================================== FAILURES ===================================
________________________ test_check_link_response_only _________________________

app = <SphinxTestApp buildername='linkcheck'>

    @pytest.mark.sphinx(
        'linkcheck', testroot='linkcheck', freshenv=True,
        confoverrides={'linkcheck_anchors': False})
    def test_check_link_response_only(app):
        with http_server(DefaultsHandler):
            app.build()
    
        # JSON output
        assert (app.outdir / 'output.json').exists()
        content = (app.outdir / 'output.json').read_text(encoding='utf8')
    
        rows = [json.loads(x) for x in content.splitlines()]
        rowsby = {row["uri"]: row for row in rows}
>       assert rowsby["http://localhost:7777/#top"]["status"] == "working"
E       AssertionError: assert 'broken' == 'working'
E         - working
E         + broken

tests/test_build_linkcheck.py:184: AssertionError
----------------------------- Captured stderr call -----------------------------
127.0.0.1 - - [05/Dec/2023 13:16:07] "HEAD / HTTP/1.1" 200 -
127.0.0.1 - - [05/Dec/2023 13:16:07] "HEAD / HTTP/1.1" 200 -
127.0.0.1 - - [05/Dec/2023 13:16:07] "HEAD /image.png HTTP/1.1" 404 -
127.0.0.1 - - [05/Dec/2023 13:16:07] "GET /image.png HTTP/1.1" 404 -
127.0.0.1 - - [05/Dec/2023 13:16:07] "HEAD /anchor.html HTTP/1.1" 200 -
127.0.0.1 - - [05/Dec/2023 13:16:07] "HEAD / HTTP/1.1" 200 -
127.0.0.1 - - [05/Dec/2023 13:16:07] "HEAD /image2.png HTTP/1.1" 404 -
127.0.0.1 - - [05/Dec/2023 13:16:07] "HEAD / HTTP/1.1" 200 -
127.0.0.1 - - [05/Dec/2023 13:16:07] "GET /image2.png HTTP/1.1" 404 -
127.0.0.1 - - [05/Dec/2023 13:16:07] "HEAD / HTTP/1.1" 200 -
----------------------------------------
Exception occurred during processing of request from ('127.0.0.1', 34022)
Traceback (most recent call last):
  File "/nix/store/480sm7i0dn11hf0lmx4wl4gk9w1k5s8r-python3-3.11.6/lib/python3.11/socketserver.py", line 691, in process_request_thread
    self.finish_request(request, client_address)
  File "/nix/store/480sm7i0dn11hf0lmx4wl4gk9w1k5s8r-python3-3.11.6/lib/python3.11/socketserver.py", line 361, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/nix/store/480sm7i0dn11hf0lmx4wl4gk9w1k5s8r-python3-3.11.6/lib/python3.11/socketserver.py", line 755, in __init__
    self.handle()
  File "/nix/store/480sm7i0dn11hf0lmx4wl4gk9w1k5s8r-python3-3.11.6/lib/python3.11/http/server.py", line 438, in handle
    self.handle_one_request()
  File "/nix/store/480sm7i0dn11hf0lmx4wl4gk9w1k5s8r-python3-3.11.6/lib/python3.11/http/server.py", line 404, in handle_one_request
    self.raw_requestline = self.rfile.readline(65537)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/480sm7i0dn11hf0lmx4wl4gk9w1k5s8r-python3-3.11.6/lib/python3.11/socket.py", line 706, in readinto
    return self._sock.recv_into(b)
           ^^^^^^^^^^^^^^^^^^^^^^^
ConnectionResetError: [Errno 104] Connection reset by peer
```

kloenk reviewed (approved):
changes LGTM

blitz commented:
@amjoseph-nixpkgs @alyssais It would be great if you have some time to look over this PR. :santa:

alyssais commented on pkgs/os-specific/linux/rust-out-of-tree-module/default.nix:
```diff
@@ -0,0 +1,28 @@
+{ lib, fetchFromGitHub, kernel }:
+kernel.stdenv.mkDerivation {
```
Why not normal stdenv?

blitz replied:
Kernel modules should be built with the stdenv that the kernel itself was built with. Otherwise, it might lead to weird problems. I think we're not very consistent about this though.

RaitoBezarius commented on nixos/tests/kernel-rust.nix:
```diff
@@ -0,0 +1,30 @@
+
+  nodes.machine = { config, pkgs, ... }:
+    {
+      boot.kernelPackages = pkgs.linuxPackages_6_6;
```
nit: we should maybe use `latest` here because 6_6 might disappear at some point, or I can count on you updating it when it starts breaking.

blitz replied:
I used 6.6 specifically, because newer versions require finding the right patches and that would block updating `latest`. Since 6.6 is a LTS version, we should be good here for a long time and I expect Rust support for `latest` to be in a much better shape as 2024 progresses.

And yes, you can count on me fixing this if it breaks. :+1:

RaitoBezarius commented on pkgs/top-level/linux-kernels.nix:
```diff
@@ -182,6 +182,8 @@ in {
       kernelPatches = [
         kernelPatches.bridge_stp_helper
         kernelPatches.request_key_helper
+        kernelPatches.rust-1-72-fix
```
nit: I'm a bit sad we need those patches even if we don't use Rust in the kernel but I guess doing conditionals here is worse in terms of rebuilds.

blitz replied:
To use the patches was the consensus in an earlier version of this PR. It was deemed less bad than having to package the matching rustc/bindgen for each kernel version. Given that they only patch Rust code, I would say it's ok.

The situation should be temporary and during 2024 all of the patching should go away.

yu-re-ka replied:
> The situation should be temporary and during 2024 all of the patching should go away.

What is the source on that?
I want this to be merged, and I still think patches are the best idea (I wrote many of the patches for nixos-apple-silicon).
But since the linux kernel has its own fork of the alloc crate vendored in and requires very fine grained control, it will always use unstable features (which will not be stabilized so soon, or ever).
Having unstable features means the code is not generally forward-compatible.
I am not aware of a plan to change this, so I would be going in with the expectation that upgrading rust in the future is always going to involve tinkering with the kernel a bit (but I'm willing to do this / have been doing this in the last months).

ojeda replied:
We want to eventually support a minimum version, but when that will actually happen is unknown -- discussions with upstream Rust are ongoing. There are some details at https://rust-for-linux.com/rust-version-policy and https://rust-for-linux.com/unstable-features (we also have a section about `alloc` there).

RaitoBezarius reviewed (approved):
I don't have as much depth as qyliss or amjoseph on all our kernel internals in nixpkgs, but this looks good to me.

alyssais commented:
Could you squash the first two commits together?  It doesn't make much sense to enable it by default, then disable it one commit later.

blitz commented:
> Could you squash the first two commits together? It doesn't make much sense to enable it by default, then disable it one commit later.

Sure will do. I just wanted to keep attribution to @yu-re-ka.

yuyuyureka committed:
linux_latest: optionally build Linux 6.7 and onwards with rust support

With the current patching situation around enabling Linux to build
with whatever rustc/bindgen version we have in nixpkgs, it feels like
supporting a Rust-by-default build at this time is not fully
justified.

Co-authored-by: Julian Stecklina <julian.stecklina@cyberus-technology.de>

blitz commented:
@alyssais I've squashed the first two commits.

alyssais commented:
> Sure will do. I just wanted to keep attribution to @yu-re-ka.

Yeah, please keep that.  You can do that with `--author` when amending the commit, and then add yourself in Co-authored-by.

blitz commented:
> > Sure will do. I just wanted to keep attribution to @yu-re-ka.
> 
> Yeah, please keep that. You can do that with `--author` when amending the commit, and then add yourself in Co-authored-by.

Done.

blitz commented:
@alyssais I've rebased this PR and limited the changes to Linux 6.7 (`linux_testing`). The kernel builds, but otherwise, it's a bit hard to test, because `staging` seems b0rken. `perl5.38.2-XML-LibXML-2.0209` which is a dependency of the test driver doesn't build. :crying_cat_face:

blitz commented:
I'm probably not going to be able to work on this for the next few weeks, because I'm traveling. 🏝️ If anyone has time to push this forwt, that would be great!

alyssais commented:
I'm happy with this, and I fixed the libxml thing, so it probably just needs a rebase.

K900 reviewed (approved):
LGTM.

RaitoBezarius commented:
> I'm happy with this, and I fixed the libxml thing, so it probably just needs a rebase.

@blitz is away from computers, I think anyone can do the rebase and @blitz would probably very happy about it too.

blitz commented:
Thanks for taking this over the finishing line! 🥳

trofi commented:
Bisect says that ec2016dda402ded1261d2af0e29d02506b191eab `doc: explain how to enable Rust support in the Linux kernel` broke `nixosTests.man` build in `staging` as:

```
$ nix build --no-link -f. nixosTests.man
...
RuntimeError: failed to render manual manual.md
error:
        failed to render manual manual.md

caused by:
        processing included file configuration/configuration.md from line 41

caused by:
        processing included file configuration/linux-kernel.chapter.md from line 15

caused by:
        heading in line 95 does not have an id
```

alyssais commented:
https://github.com/NixOS/nixpkgs/pull/281181

RaitoBezarius commented:
> Why did this go to staging? I can see the rebuild count according to ofborg is _rather_ high, but rebuilding my own machine with it makes almost 0 rebuilds (so it doesn't affect linux_headers for example).

I think you answered yourself :P — the ofborg count is usually the thing we use for decisionmaking. The fact that it's faster to rebuild is a good thing but requires proving it before sending this to master and increase infra people work, I'd say. 

Is there any reason that waiting for staging is problematic on this one?

samueldr commented:
This changeset, and specifically 8a5f6594daa32af3686d13eb2efa5092598f9bbe was bisected as a regression that prevents kernel patches to be conditionally applied depending on `config.boot.kernelPackages.kernel.version` in a NixOS configuration.

Minified reproduction (excuse the meaninglessness of the actual configured system):

```nix
# configuration.nix
{ config, lib, ... }:

let
  patch = {
    patch = ""; # this can be ignored, it's an eval-time failure
  };
  selectPatch = kernel:
    let
      atLeast = lib.versionAtLeast kernel.version;
    in

    if atLeast "0.0" then # value can be ignored here
      patch
    else
      patch
  ;
in
{
  boot.kernelPatches = [
    (selectPatch config.boot.kernelPackages.kernel)
  ];

  # This can be ignored
  fileSystems = { "/" = { device = "/dev/null"; }; };
  boot.loader.grub.enable = false;
}
```

The root cause of the issue comes from `withRust` being used in the `kernel`'s `config` attribute, and `withRust` depending on the *parsed* final configuration, which in turn, depends on the patches and the fact they can add to the kernel configuration.

Since `config.boot.kernelPackages.kernel` now depends on `configfile` being parsed, which depends on `config.boot.kernelPatches`, we get a loop.

* * *

I don't have any suggestion at the time being. I haven't had the energy to deeply think about the possible solutions. Though I thought I should report the issue.

(I've just now checked staging and master, still fails the same.)

* * *

<details><summary>Bisect log</summary>

```
~/.../nixpkgs $ git bisect log
git bisect start
# status: waiting for both good and bad commits
# bad: [e92b6015881907e698782c77641aa49298330223] Merge pull request #285831 from r-ryantm/auto-update/python311Packages.qdrant-client
git bisect bad e92b6015881907e698782c77641aa49298330223
# status: waiting for good commit(s), bad commit known
# good: [bd645e8668ec6612439a9ee7e71f7eac4099d4f6] Merge pull request #278149 from r-ryantm/auto-update/mimir
git bisect good bd645e8668ec6612439a9ee7e71f7eac4099d4f6
# good: [bbd92ae0479d169c7917a92e482d598e6ea6091d] Merge pull request #280561 from RaitoBezarius/fix-listmonk-module
git bisect good bbd92ae0479d169c7917a92e482d598e6ea6091d
# good: [39a2e0bb0390869e981586d53a2a97bf9673713b] Merge pull request #283396 from newAM/clevis-clarify
git bisect good 39a2e0bb0390869e981586d53a2a97bf9673713b
# bad: [8ac7d981cafdc7c598727d591afab07388bf76f4] Merge pull request #281426 from SuperSandro2000/celeray-xdist
git bisect bad 8ac7d981cafdc7c598727d591afab07388bf76f4
# bad: [c61980d92bbde58bdbe10a4753c208114c33d1f7] coqPackages.flocq: 4.1.3 -> 4.1.4
git bisect bad c61980d92bbde58bdbe10a4753c208114c33d1f7
# bad: [359064396f556da8552ce3437668bd6b732894d5] vimPlugins.haskell-scope-highlighting-nvim: init at 2023-04-29
git bisect bad 359064396f556da8552ce3437668bd6b732894d5
# bad: [35f426a31bd783bfb5a66a5e129d472344db118b] Merge staging-next into staging
git bisect bad 35f426a31bd783bfb5a66a5e129d472344db118b
# good: [43d0e3ff336309ccc7d6f8eac0b5e4c5b885b78b] Merge pull request #253133 from jiegec/libunwind
git bisect good 43d0e3ff336309ccc7d6f8eac0b5e4c5b885b78b
# bad: [57a3e97b68a5b1de6e975023866c988ce8a3466f] Merge pull request #280144 from mweinelt/jinja-3.1.3
git bisect bad 57a3e97b68a5b1de6e975023866c988ce8a3466f
# good: [f1c4794baf431288cc065adfee3dde047d3a005f] Merge pull request #276595 from r-ryantm/auto-update/vim
git bisect good f1c4794baf431288cc065adfee3dde047d3a005f
# good: [d6760d87988a9874f9877e00ecae0078cc217810] Merge pull request #277321 from atorres1985-contrib/bluez_staging
git bisect good d6760d87988a9874f9877e00ecae0078cc217810
# bad: [32ccbe0927fc910f8fa3541f27f1aeb0effd8fbd] Merge pull request #272029 from blitz/rust-for-linux
git bisect bad 32ccbe0927fc910f8fa3541f27f1aeb0effd8fbd
# good: [243e5f6f58048a2f99d1311166c608ebaf46479f] Merge pull request #280724 from K900/kf5-5.114
git bisect good 243e5f6f58048a2f99d1311166c608ebaf46479f
# bad: [5fca7feab9b1811348b6bb86fc0bbfa86a3acfea] rust-out-of-tree-module: init
git bisect bad 5fca7feab9b1811348b6bb86fc0bbfa86a3acfea
# bad: [8a5f6594daa32af3686d13eb2efa5092598f9bbe] linux_latest: optionally build Linux 6.7 and onwards with rust support
git bisect bad 8a5f6594daa32af3686d13eb2efa5092598f9bbe
# first bad commit: [8a5f6594daa32af3686d13eb2efa5092598f9bbe] linux_latest: optionally build Linux 6.7 and onwards with rust support
```

</details>
